pkg/erofs/erofs.go

```diff
@@ -0,0 +1,729 @@
+// [ (block 2) dirent 4 | dirent 5 | name 4 | name 5 | optional padding ]
+// ...
+// [ (block N) dirent M | dirent M+1 | name M | name M+1 | optional padding ]
+//
+// [ (metadata block) inode | optional fields | dirent M+2 | dirent M+3 | name M+2 | name M+3 | optional padding ]
+//
+// All directory entries (except the first two: "." and "..") are _strictly_ recorded in alphabetical order.
+func (i *Inode) IterDirents(cb func(name string, typ uint8, nid uint64) error) error {
+	if !i.IsDir() {
+		return linuxerr.ENOTDIR
+	}
+
+	blocks := i.blocks()
+	blockSize := i.image.BlockSize()
+
+	start := i.dataOff
+	if blocks == 1 && i.idataOff != 0 {
+		start = i.idataOff
+	}
+
+	// Iterate all the blocks which contain dirents.
+	for blocks > 0 {
+		// Unmarshal the first dirent in the current block.
+		direntOff := start
+		d := &Dirent{}
+		if err := i.image.UnmarshalAt(d, direntOff); err != nil {
+			return err
+		}
+		// Apart from the offset of the first filename, nameOff0 also indicates
+		// the total number of dirents in this block.
+		nameOff0 := start + uint64(d.NameOff)
+		maxSize := blockSize
+		if blocks == 1 {
+			if tailSize := uint32(i.size) & (blockSize - 1); tailSize != 0 {
+				maxSize = tailSize
+			}
+		}
+		// Iterate all the dirents in this block.
+		for d != nil {
+			var (
+				next    *Dirent
+				nameLen uint32
+			)
+			direntOff += uint64(d.SizeBytes())
+			lastDirent := direntOff >= nameOff0
+			if lastDirent {
+				// There is no more dirent in this block, d is the last one.
+				next = nil
+				nameLen = maxSize - uint32(d.NameOff)
+			} else {
+				// Unmarshal the next adjacent dirent.
+				next = &Dirent{}
+				if err := i.image.UnmarshalAt(next, direntOff); err != nil {
+					return err
+				}
+				nameLen = uint32(next.NameOff - d.NameOff)
+			}
+
+			buf, err := i.image.BytesAt(start+uint64(d.NameOff), uint64(nameLen))
+			if err != nil {
+				return err
+			}
+			if lastDirent {
+				if n := bytes.IndexByte(buf, 0); n != -1 {
+					nameLen = uint32(n)
+				}
+			}
+			if nameLen > MaxNameLen {
+				log.Warningf("Corrupted dirent at inode (nid=%v)", i.Nid())
+				return linuxerr.EUCLEAN
+			}
+			name := string(buf[:nameLen])
+			if err := cb(name, d.FileType, d.Nid); err != nil {
+				return err
+			}
+
+			d = next
+		}
+
+		blocks--
+
+		if blocks == 1 && i.idataOff != 0 {
+			// If we have any inline data and this is the last block, we need to process it now.
+			start = i.idataOff
+		} else {
+			// Just go on to process the next adjacent block.
+			start += uint64(blockSize)
+		}
+	}
+	return nil
+}
+
+// Readlink reads the link target.
+func (i *Inode) Readlink() (string, error) {
+	if !i.IsSymlink() {
+		return "", linuxerr.EINVAL
+	}
+	off := i.dataOff
+	if i.idataOff != 0 {
+		// Inline symlink data shouldn't cross block boundary.
+		if i.blocks() > 1 {
+			log.Warningf("Inline data cross block boundary at inode (nid=%v)", i.Nid())
+			return "", linuxerr.EUCLEAN
+		}
```

Comment:
You are right. In Linux, it will do this check when inline data is present:

https://github.com/torvalds/linux/blob/6465e260f48790807eef06b583b38ca9789b6072/fs/erofs/inode.c#L223-L230

And when inline data is not present, it will truncate it to page size:

https://github.com/torvalds/linux/blob/6465e260f48790807eef06b583b38ca9789b6072/fs/erofs/inode.c#L211-L215

https://github.com/torvalds/linux/blob/6465e260f48790807eef06b583b38ca9789b6072/fs/namei.c#L5167

I will follow the same behaviour in the next version.